internal/pkg/service/common/configpatch/dump.go

```diff
@@ -0,0 +1,174 @@
+package configpatch
+
+import (
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/keboola/go-utils/pkg/orderedmap"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/configmap"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/strhelper"
+)
+
+// DumpKV is result of the patch operation for a one configuration key.
+type DumpKV struct {
+	// KeyPath is a configuration key, parts are joined with a dot "." separator.
+	KeyPath string `json:"key"`
+	// Value is an actual value of the configuration key.
+	Value any `json:"value"`
+	// DefaultValue of the configuration key.
+	DefaultValue any `json:"defaultValue"`
+	// Overwritten is true, if the DefaultValue was replaced by a value from the path.
+	Overwritten bool `json:"overwritten"`
+	// Protected configuration keys can be modified only by a super-admin.
+	Protected bool `json:"protected"`
+	// Validation contains validation rules of the field, if any.
+	Validation string `json:"validation,omitempty"`
+}
+
+// DumpKVs generates key-value pairs from a configuration structure and a patch structure.
+// Only keys found in both, configuration and patch structure, are processed.
+// The structure is flattened, keys are joined with a dot "." separator.
+// Each key-value pair contains information whether the value was overwritten from the patch or not.
+func DumpKVs(configStruct, patchStruct any, opts ...Option) ([]DumpKV, error) {
```

Comment:
This new function generates key-value pairs, see real world example, in the PR description.